### DISCUSSION TITLE
ACP-25 VM Application Errors

### DISCUSSION THREAD

**Discussion by joshua-kim** [avalanche-foundation/ACPs#26] [Brainstorming]
Support VM defined application errors. Introduces a new p2p message type. https://github.com/avalanche-foundation/ACPs/pull/25

**Comment by joshua-kim** [comment]
1. I think `AppError` sounds better than `AppRequestFailed` but `AppRequestFailed` is the current name of the VM interface's function that's invoked in timeout. We could consider using the `AppError` name instead and just rename the VM signature.
2. Some follow-up ideas could be to score VMs, and to incentivize sending `AppRequestFailed` messages by deducting more score for network timeouts as opposed to an application defined error.

**Reply by patrick-ogrady** [reply]
1) I agree. AppRequestFailed sounds like a class of errors (could be timeout or explicit error).
2) I think this should be left to the VM to determine based on the content of the message. Some messages would be OK to fail IMO.

**Reply by joshua-kim** [reply]
Rename PR: https://github.com/ava-labs/avalanchego/pull/2321

**Comment by patrick-ogrady** [comment]
Short circuit timeout-based errors wherever we can!

**Comment by avalanche-foundation-admin** [comment]
This conversation has migrated to the official ACP discussion: https://github.com/avalanche-foundation/ACPs/discussions/39
